cou/apps/auxiliary.py

```diff
@@ -457,3 +462,146 @@ async def _verify_nova_compute(self, target: OpenStackRelease) -> None:
             raise ApplicationError(
                 f"Units '{', '.join(units_not_upgraded)}' did not reach {target}."
             )
+
+
+@AppFactory.register_application(["vault"])
+class Vault(AuxiliaryApplication):
+    """Application for vault."""
+
+    wait_timeout = LONG_IDLE_TIMEOUT
+    wait_for_model = True
+
+    def _get_cacert_file(self) -> Optional[str]:
+        """Read cert file and write into temporary file.
+
+        :return: Temporary file path
+        :rtype: str
+        """
+        cacert_b64 = self.config["ssl-ca"].get("vault")
+        cacert_file = None
+        if cacert_b64:
+            with tempfile.NamedTemporaryFile(mode="wb", delete=False) as fp:
+                fp.write(base64.b64decode(cacert_b64))
+                cacert_file = fp.name
+        return cacert_file
+
+    async def _get_unit_api_url(self, unit_name: str) -> str:
+        """Get unit's vault api address.
+
+        :param unit_name: vault unit name
+        :type unit_name: str
+        :return: unit's vault api url
+        :rtype: str
+        """
+        # Use vip as address if vip is being used.
+        if self.config["vip"].get("value"):
+            address = self.config["vip"].get("value")
+        else:
+            juju_unit = await self.model.get_unit(unit_name)
+            address = juju_unit.public_address
+
+        transport = "http"
+        if self.config["ssl-cert"].get("value"):
```

Comment:
If it's possible to have non-TLS enabled vault, does that create a potential security issue? Should we simply let the user to unseal itself?

Reply:
What behavior you want? We stop unseal because TLS is required?